### PR TITLE
[Image recs] Add basic Image Recommendations API calls

### PR DESCRIPTION
**Pull request by mazevedofs**
**Phabricator:**  https://phabricator.wikimedia.org/T355321

### Notes
* This PR adds the basic calls to retrieve image recommendation tasks and the recommendations per article

### Test Steps
1. Cherry-pick this commit to perform the calls on the explore feed e2134195404fd4e025ba7d7354915314ec05bba4
2. Make sure you either see the objects on the debug prints or you can inspect the calls on a proxy debugger

### CODE REVIEW

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    let project: WKProject
+
+    public init (project: WKProject) {
+        self.project = project
+    }
+
+    // MARK: GET Methods
+
+    // TODO: to get more tasks, send used IDS to `ggtexcludepageids` parameter
+    // TODO: remove debug prints
```
Comment (mazevedofs) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:15]: I'm keeping the debug prints for a bit, LMK if you're against it 😅

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 2]
```diff
@@ -0,0 +1,157 @@
+import Foundation
+
+public final class WKGrowthTasksDataController {
+
+    var service = WKDataEnvironment.current.mediaWikiService
```
Comment (tonisevener) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:5]: Maybe make this `private let` if you don't expect to use or change it outside of this class.

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    // MARK: GET Methods
+
+    // TODO: to get more tasks, send used IDS to `ggtexcludepageids` parameter
+    // TODO: remove debug prints
+    public func getGrowthAPITask(task: WKGrowthTaskType = .imageRecommendation, completion: @escaping (Result<[WKGrowthTask.Page], Error>) -> Void) {
+
+        guard let service else {
+            return
```
Comment (tonisevener) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:19]: It'd be good to call the error handler for any early returns. I've got some more generalized ones in `WKDataControllerError`, so you can use `completion(.failure(WKDataControllerError.mediaWikiServiceUnavailable)` here.

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        let parameters = [ "action": "query",
+                           "generator": "growthtasks",
+                           "formatversion": "2",
+                           "format": "json",
+                           "ggttasktypes": "\(task.rawValue)",
+                           "ggtlimit": "10"
+        ]
+
+        guard let url = URL.mediaWikiAPIURL(project: project) else {
+            return
```
Comment (tonisevener) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:32]: You can call `completion(.failure(WKDataControllerError.failureCreatingRequestURL)` before this return.

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        }
+    }
+
+    public func getImageSuggestionData(pageIDs: [String], completion: @escaping (Result<[WKImageRecommendation.Page], Error>) -> Void) {
+
+        let pipeEncodedPageIds = pageIDs.joined(separator: "|")
+        var recommendationsPerPage:[WKImageRecommendation.Page] = []
+
+        guard let service else {
+            return
```
Comment (tonisevener) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:55]: Same as above - call error handler.

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+
+        let parameters = [ "action": "query",
+                           "formatversion": "2",
+                           "format": "json",
+                           "prop":"growthimagesuggestiondata",
+                           "pageids" : pipeEncodedPageIds
+        ]
+
+        guard let url = URL.mediaWikiAPIURL(project: project) else {
+            return
```
Comment (tonisevener) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:66]: Same as above - call error handler.

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        }
+
+        let request = WKMediaWikiServiceRequest(url: url, method: .GET, parameters: parameters)
+        service.performDecodableGET(request: request) { (result: Result<WKImageRecommendationAPIResponse, Error>) in
+
+            switch result {
+            case .success(let response):
+                print(response)
+                recommendationsPerPage.append(contentsOf: self.getImageSuggestions(from: response))
```
Comment (tonisevener) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:75]: Needs the completion handler called here.

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+        let request = WKMediaWikiServiceRequest(url: url, method: .GET, parameters: parameters)
+        service.performDecodableGET(request: request) { (result: Result<WKGrowthTaskAPIResponse, Error>) in
+
+            switch result {
+            case .success(let response):
+                print(response)
+                pages.append(contentsOf: self.getTaskPages(from: response))
+                completion(.success(pages))
+            case .failure(let error):
+                print(error)
```
Comment (tonisevener) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:44]: Wasn't sure if you were going to tackle failures later but for now you can pass along the error here:

`completion(.failure(error))`

File: WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+        let request = WKMediaWikiServiceRequest(url: url, method: .GET, parameters: parameters)
+        service.performDecodableGET(request: request) { (result: Result<WKImageRecommendationAPIResponse, Error>) in
+
+            switch result {
+            case .success(let response):
+                print(response)
+                recommendationsPerPage.append(contentsOf: self.getImageSuggestions(from: response))
+            case .failure(let error):
+                print(error)
```
Comment (tonisevener) [inline WKData/Sources/WKData/Data Controllers/WKGrowthTasks/WKGrowthTasksDataController.swift:77]: You can pass along the error here:

`completion(.failure(error))`

**Review by tonisevener — COMMENTED** [review]
Looks good! Just minor things - feel free to ignore if you plan on tackling them later.

**Pull request merged**